server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -438,17 +488,25 @@ public void invariant() {
                 assert prevotingRound == null : prevotingRound;
                 assert becomingMaster || getStateForMasterService().nodes().getMasterNodeId() != null : getStateForMasterService();
                 assert leaderCheckScheduler == null : leaderCheckScheduler;
-
-                final Set<DiscoveryNode> knownFollowers = followersChecker.getKnownFollowers();
-                final Set<DiscoveryNode> lastPublishedNodes = new HashSet<>();
-                if (becomingMaster == false || publicationInProgress()) {
-                    final ClusterState lastPublishedState
-                        = currentPublication.map(Publication::publishedState).orElse(coordinationState.get().getLastAcceptedState());
+                assert applierState.nodes().getMasterNodeId() == null || getLocalNode().equals(applierState.nodes().getMasterNode());
+
+                if (becomingMaster && activePublicationInProgress() == false) {
+                    // cluster state update task to become master is submitted to MasterService, but publication has not started yet
+                    assert followersChecker.getKnownFollowers().isEmpty() : followersChecker.getKnownFollowers();
+                } else {
+                    final ClusterState lastPublishedState;
+                    if (activePublicationInProgress()) {
+                        // active publication in progress: followersChecker is up-to-date with nodes that we're actively publishing to
+                        lastPublishedState = currentPublication.get().publishedState();
+                    } else {
+                        // no active publication: followersChecker is up-to-date with the nodes of the latest publication
+                        lastPublishedState = coordinationState.get().getLastAcceptedState();
+                    }
+                    final Set<DiscoveryNode> lastPublishedNodes = new HashSet<>();
                     lastPublishedState.nodes().forEach(lastPublishedNodes::add);
-                    assert lastPublishedNodes.remove(getLocalNode());
-                    assert lastPublishedNodes.equals(knownFollowers) : lastPublishedNodes + " != " + knownFollowers
-                        + " [becomingMaster=" + becomingMaster + ", publicationInProgress=" + publicationInProgress() + "]";
-                    // TODO instead assert that knownFollowers is updated appropriately at the end of each publication
+                    assert lastPublishedNodes.remove(getLocalNode()); // followersChecker excludes local node
+                    assert lastPublishedNodes.equals(followersChecker.getKnownFollowers()) :
+                        lastPublishedNodes + " != " + followersChecker.getKnownFollowers();
```

Comment:
This assertion fails:

```:server:test -Dtests.class=org.elasticsearch.cluster.coordination.CoordinatorTests -Dtests.method=testUnresponsiveLeaderDetectedEventually -Dtests.seed=B39FA830E397231C:E88D9536D29A794 -Dtests.jvm.argline=-Dhppc.bitmixer=DETERMINISTIC```

(I have some local modifications so may not reproduce exactly)

The reason is the same as in https://github.com/elastic/elasticsearch/pull/34257/files#r222583570: a publication is committed but times out before being accepted locally. Here it's a commit that's adding a node after the cluster formed, so the known followers are updated at the start of the publication but the last-accepted state doesn't reflect this.